zkevm-circuits/src/bytecode_circuit/bytecode_unroller.rs

```diff
@@ -0,0 +1,880 @@
+        // q_enable
+        if enable {
+            self.q_enable.enable(region, offset)?;
+        }
+
+        // q_first
+        region.assign_fixed(
+            || format!("assign q_first {}", offset),
+            self.q_first,
+            offset,
+            || Ok(F::from((offset == 0) as u64)),
+        )?;
+
+        // q_last
+        if last {
+            self.q_last.enable(region, offset)?;
+        }
+
+        // Advices
+        for (name, column, value) in &[
+            ("hash", self.hash, hash),
+            ("index", self.index, index),
+            ("is_code", self.is_code, is_code),
+            ("byte", self.byte, byte),
+            ("push_rindex", self.push_rindex, F::from(push_rindex)),
+            ("hash_rlc", self.hash_rlc, hash_rlc),
+            ("hash_length", self.hash_length, hash_length),
+            ("byte_push_size", self.byte_push_size, byte_push_size),
+            ("is_final", self.is_final, F::from(is_final as u64)),
+            ("padding", self.padding, F::from(padding as u64)),
+        ] {
+            region.assign_advice(
+                || format!("assign {} {}", name, offset),
+                *column,
+                offset,
+                || Ok(*value),
+            )?;
+        }
+
+        // push_rindex_is_zero_chip
+        push_rindex_is_zero_chip.assign(
+            region,
+            offset,
+            Some(push_rindex_prev),
+        )?;
+
+        Ok(())
+    }
+
+    pub(crate) fn load(
+        &self,
+        layouter: &mut impl Layouter<F>,
+        bytecodes: &[UnrolledBytecode<F>],
+    ) -> Result<(), Error> {
+        // push table: BYTE -> NUM_PUSHED:
+        // [0, OpcodeId::PUSH1[ -> 0
+        // [OpcodeId::PUSH1, OpcodeId::PUSH32] -> [1..32]
+        // ]OpcodeId::PUSH32, 256[ -> 0
+        layouter.assign_region(
+            || "push table",
+            |mut region| {
+                for byte in 0usize..256 {
+                    let push_size = get_push_size(byte as u8);
+                    for (name, column, value) in &[
+                        ("byte", self.push_table[0], byte as u64),
+                        ("push_size", self.push_table[1], push_size),
+                    ] {
+                        region.assign_fixed(
+                            || format!("Push table assign {} {}", name, byte),
+                            *column,
+                            byte,
+                            || Ok(F::from(*value)),
+                        )?;
+                    }
+                }
+                Ok(())
+            },
+        )?;
+
+        // keccak table
+        layouter.assign_region(
+            || "keccak table",
+            |mut region| {
+                for (offset, bytecode) in
+                    bytecodes.iter().map(|v| v.bytes.clone()).enumerate()
+                {
+                    let hash: F = keccak(&bytecode[..], self.r);
+                    let rlc: F = linear_combine(bytecode.clone(), self.r);
```

Comment:
one more:  keccak table validity(hash correctness) will be checked by elsewhere like RW_table, Tx_table, right?

Reply:
That's right, will be checked in the keccak circuit.